File: 1697-checking-existence-of-edge-length-limited-paths/1697-checking-existence-of-edge-length-limited-paths.js
/**
 * @param {number} n
 * @param {number[][]} edgeList
 * @param {number[][]} queries
 * @return {boolean[]}
 */
var distanceLimitedPathsExist = function(n, edgeList, queries) {
  edgeList.sort((a, b) => a[2] - b[2]);
  queries = queries.map((query, i) => [i, ...query]).sort((a, b) => a[3] - b[3]);
  const uf = new UnionFind(n);
  const ans = new Array(queries.length);
  let j = 0;
  
  for (let [i, p, q, limit] of queries) {
    while (j < edgeList.length && edgeList[j][2] < limit) {
      const [u, v] = edgeList[j++];
      uf.union(u, v);
    }
    
    ans[i] = uf.isConnected(p, q);
  }
  
  return ans;
};

class UnionFind {
  constructor(n) {
    this.parent = Array.from({ length: n }, (_, i) => i);
  }
  
  find(x) {
    let parent = this.parent;
    while (x !== parent[x]) {
      parent[x] = parent[parent[x]];
      x = parent[x];
    }
    
    return x;
  }
  
  isConnected(x, y) {
    return this.find(x) === this.find(y);
  }
  
  union(x, y) {
    let rootX = this.find(x);
    let rootY = this.find(y);
    if (rootX === rootY) return;
    
    this.parent[rootX] = rootY;
  }
}

File: 1697-checking-existence-of-edge-length-limited-paths/1697-checking-existence-of-edge-length-limited-paths.py
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        
    def find(self, x):
        if x != self.parent[x]:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
      
    def isConnected(self, x, y):
        return self.find(x) == self.find(y)
      
    def union(self, x, y):
        root_x, root_y = self.find(x), self.find(y)
        if root_x == root_y:
            return 
        self.parent[root_x] = root_y
        

class Solution:
    def distanceLimitedPathsExist(self, n: int, edgeList: List[List[int]], queries: List[List[int]]) -> List[bool]:
        edgeList.sort(key=lambda x: x[2])
        queries = sorted([[i] + q for i, q in enumerate(queries)], key=lambda x: x[3])
        uf = UnionFind(n)
        ans = [False] * len(queries)
        j = 0
        
        for i, p, q, limit in queries:
            while j < len(edgeList) and edgeList[j][2] < limit:
                u, v, _ = edgeList[j]
                uf.union(u, v)
                j += 1
            ans[i] = uf.isConnected(p, q)
        return ans